### PR TITLE
More docs+examples (freelist, pyany, pymodule, puthon::check_signals)

### PR DESCRIPTION
**Pull request by mejrs**
Adds some documentation.

Thanks to @birkenfeld for explaining some things around check_signals ☺️

<details><summary>Freelist</summary>
<p>

image

</p>
</details>

<details><summary>PyAny</summary>
<p>

image
image
image
image
image
image

</p>
</details>

<details><summary>PyModule</summary>
<p>

image
image
image
image
image
image
image

</p>
</details>

<details><summary>Python::check_signals</summary>
<p>

image

</p>
</details>

### CODE REVIEW

File: src/types/module.rs
Comment (mejrs) [inline src/types/module.rs:179]: I turned this into an actual `#[deprecated(...)]`.

I'm assuming this is still what we want...?

**Review by birkenfeld — CHANGES_REQUESTED** [review]

File: src/types/any.rs
Comment (davidhewitt) [reply src/types/any.rs:214]: (I tend to do a newline above but not below)

File: src/types/module.rs
Comment (davidhewitt) [reply src/types/module.rs:185]: Agreed with attribute!

File: src/python.rs
Comment (mejrs) [reply src/python.rs:564]: Oops 🙊

File: src/types/module.rs
Comment (mejrs) [reply src/types/module.rs:224]: This one tripped me up, which is why I specifically included it  👀

File: src/types/any.rs
Comment (mejrs) [reply src/types/any.rs:401]: :)

File: src/types/module.rs
Comment (mejrs) [reply src/types/module.rs:76]: I hope so too.

There are some efforts to officially include something like this in Rustdoc, but unfortunately it seems to be stuck in limbo wrt. figuring out syntax that'll play nice with the Markdown format.

File: src/types/module.rs
Comment (mejrs) [reply src/types/module.rs:279]: Well it does create a module hierarchy, but it doesn't (currently) act like a package - which is why I avoided using the word "package". In hindsight that's probably a lie by omission.

But it acting like a package is probably what people *expect* when they add submodules. What are your thoughts on adding something like this below it?

"Note that pyo3 does not yet support using modules as packages; see https://github.com/PyO3/pyo3/issues/1517#issuecomment-808664021 for a hack to make that work"

Comment (birkenfeld) [reply src/types/module.rs:279]: I'm not sure it's actually *possible* to support it other than with what I posted there. As far as CPython is concerned, every module is a file, and since the compiled extension is also a file it is also treated as a single module (e.g. only one `PyInit_modname` function is used).

At the time I was looking but couldn't easily find C extensions that create a package hierarchy; I'm guessing that with C, the desire to write whole sets of modules is not as big as with Rust :D  With C, it is of course also more straightforward to provide lots of small `.so` libs - for Rust they'd have to be in separate crates.

File: src/types/any.rs
Comment (mejrs) [reply src/types/any.rs:214]: Agreed, I'll do that.

File: src/python.rs
Comment (mejrs) [reply src/python.rs:559]: I've decided to change this more towards the original wording:
```
Lets the Python interpreter check and handle any pending signals and invoke the
corresponding signal handlers.

Returns `Err(PyErr)` if any signal handler raises an exception.
```

File: src/python.rs
Comment (mejrs) [reply src/python.rs:579]: 👍

File: src/types/module.rs
Comment (mejrs) [reply src/types/module.rs:185]: 👍

File: src/types/module.rs
Comment (mejrs) [reply src/types/module.rs:158]: 👍

File: src/types/any.rs
Comment (mejrs) [reply src/types/any.rs:349]: I'm not sure..but you're right, I'll revert that.

File: src/types/module.rs
Comment (mejrs) [reply src/types/module.rs:279]: What about this?
```
Adds a submodule to a module.

This is especially useful for creating module hierarchies.

Note that this doesn’t define a package, so this won’t allow Python code to directly import submodules like from my_module import submodule. For more information, see #759 and #1517.
```

File: src/types/module.rs
Comment (mejrs) [reply src/types/module.rs:179]: I've removed the deprecation

File: src/freelist.rs
Comment (davidhewitt) [inline src/freelist.rs:20]: It's perhaps worth adding a note that it's really not expected for the user to ever interact with this trait. As far as I'm concerned, it's an implementation detail that's very likely to need to be rewritten when we optimize pyclass creation. (We'll preserve `#[pyclass(freelist = N)]` but not necessarily any of the mechanics.)

File: src/python.rs
Comment (davidhewitt) [inline src/python.rs:560]: There's quite a few "and" in this sentence now, maybe the below is slightly clearer?

```suggestion
    /// Lets the Python interpreter check and handle any pending signals. This will invoke the
    /// corresponding signal handlers registered in Python (if any).
```

File: src/python.rs
Comment (davidhewitt) [inline src/python.rs:566]: I tend to avoid using "you" and "we" to refer to the reader and us as implementers, and instead just describe the properties of the code:

```suggestion
    /// `KeyboardInterrupt`. For this reason it is good practice to call this function regularly
    /// as part of long-running Rust functions so that users can cancel it.
```

File: src/python.rs
Comment (davidhewitt) [inline src/python.rs:578]: Similarly use of "we":

```suggestion
    ///         // As this loop is infinite it should check for signals every once in a while.
    ///         // Using `?` causes any `PyErr` (potentially containing `KeyboardInterrupt`) to break out of the loop.
```

File: src/python.rs
Comment (davidhewitt) [inline src/python.rs:582]: Does rustdoc handle this correctly, or does the `#` need to go further left?

```suggestion
    /// #         break Ok(()) // don't actually loop forever
```

File: src/python.rs
Comment (davidhewitt) [inline src/python.rs:580]:
```suggestion
    ///
```

**Review by davidhewitt — APPROVED** [review]
This is brilliant, thanks @mejrs for writing and @birkenfeld for a lot of reviewing over the past couple days!

I have a few small suggestions, and then I'm enthusiastic to merge this unless there's any other feedback or questions needed.

File: src/python.rs
Comment (mejrs) [reply src/python.rs:582]: This is correctly handled as far as I can tell. It doesn't show up in documentation and terminates correctly in doctests.

File: src/python.rs
Comment (mejrs) [reply src/python.rs:572]: Oh, nice.

Comment (mejrs) [reply src/python.rs:572]: Actually, I'd prefer to leave them in, because the invisible `main` function is copied when using the book's clipboard button, as well as by the "Run" button (for crates with a playground link). This saves users from having to manually put in a main function.

File: src/types/any.rs
Comment (mejrs) [reply src/types/any.rs:238]: I think this risks suggesting that this only happens with classes that don't define constructors. So I'd add as well that this is just *one* example of where this can happen.

There are other instances where it can happen, for example classes can sabotage their call slot:

```python
class Foo:
    __call__ = 42
```

File: src/types/module.rs
Comment (birkenfeld) [reply src/types/module.rs:297]: Sounds good for now.

File: src/python.rs
Comment (birkenfeld) [reply src/python.rs:572]: Is pyo3 even usable in the playground?!

File: src/types/any.rs
Comment (birkenfeld) [reply src/types/any.rs:238]: In other words: you can only check if you can call it, but not if you can call it successfully, with which arguments or in which environment.

This is why this check is only useful for such dynamically typed APIs with decisions based on "what kind of argument am I given here".

File: src/types/any.rs
Comment (mejrs) [reply src/types/any.rs:460]: I'm now leaning towards replacing the entire example. It's not *about* bytes literals after all, and they really should be explained in depth first. 

I'll replace it with this, I think:
```python
my_list = [1,3,4]
my_list.insert(1,2)
```

File: src/types/any.rs
Comment (mejrs) [reply src/types/any.rs:238]: > All these fine details are in the end irrelevant.

I agree. They are far too specific anyway. I've gutted it but I would like to keep the part about handling calling errors, since that seems to be what people want to do when they reach for `callable()`.

What are your thoughts on this?
image

File: src/freelist.rs
Comment (mejrs) [reply src/freelist.rs:20]: Done.

File: src/types/any.rs
Comment (birkenfeld) [reply src/types/any.rs:238]: Better, though I still maintain that the "don't use `unwrap()`" advice applies to *all* Python API operations, since most of them can raise arbitrary exceptions, and feels out of place here.

File: src/python.rs
Comment (birkenfeld) [reply src/python.rs:572]: :+1:

Comment (birkenfeld) [reply src/python.rs:572]: > But the playground link can be customized, so in principle we could host our own playground. (We'd need to host a python interpreter too? How does that work?)

Interesting question - at least a libpython would need to be available. AFAIU the playground deployment is container based, so the used container can hopefully be easily amended.

But I don't know who would host it...

**Comment by davidhewitt** [comment]
I've just taken another read, looks brilliant! Thanks very much for the massive improvement to the documentation!

**Pull request merged**
